### DISCUSSION TITLE
ILI9341 display with XPT2046 wrong touch position

### DISCUSSION THREAD

**Discussion by Jardland** [arendst/Tasmota#15792] [Q&A] [answered]
flashed to ESP8266 
compiled with
#define USE_SPI
#define USE_DISPLAY
#define USE_TOUCH_BUTTONS 
#define USE_XPT2046

DisplayMode 0
DisplayType 4
DisplayRotate 0

displaytext[zb256:0:0:60:60:2:11:4:2:B1:b257:60:0:60:60:2:11:4:2:B2:b258:120:0:60:60:2:11:4:2:B3:b259:180:0:60:60:2:11:4:2:B4:b260:0:260:60:60:2:11:4:2:B5:b261:60:260:60:60:2:11:4:2:B6:b262:120:260:60:60:2:11:4:2:B7:b263:180:260:60:60:2:11:4:2:B8:]

if you touch button 1 it changes button 4
if you touch button 8 it changes button 5

from log
touch button right top
15:30:58.159 TS : released x=35 y=28 (raw x=3587 y=3371)
touch button left bottom
15:31:41.730 TS : released x=214 y=281 (raw x=651 y=738)

How to get it to change correct button?

**Comment by gemu2015** [comment]
resistive touch must be calibrated. calibration is only included in universal display driver.  (driver 17)
this driver needs a display descriptor and you can define touch calibration data in that descriptor.
this driver is most useful with a file system on 4M esp8266.
on 1M esp8266 devices you would need to define the descriptor in user_config_override.

**Reply by Jardland** [reply]
I tried driver 17 but I could not get touch buttons to work. I tried to look into what is the problem, the y value from touch screen seems inverted so I tried to change line 188 in driver 55 to   "return 4095 - pLoc.y;//  return pLoc.y;" and now it is working OK for DisplayRotate 0 and 2. But for DisplayRotate 1 and 3 touch buttons looks OK but touch area get wider and lower. The problem with touch area is even before changing the driver.

**Comment by gemu2015** [comment, marked as answer]
you do not need to modify any code
everything is done with descriptor entries
however i realized that display doku is outdated
i will update display docs in a week or so
currently in vacation with very limited internet access

**Reply by gemu2015** [reply]
i updated display docs by missing mapping entry of display descriptor.
:M,X1,X2,Y1,Y2

in rotation 0 click into all 4 corners and note raw touch values.
then enter left upper x into X1, right upper x into X2
then enter left upper y into Y1, left lower y into Y2

the touch rotation conversions are handled by rotation table entries 
:R
4. entry contains pseudo opcode for touch rotation, see docs
